### PR TITLE
Fix memory leak when removing layers from ol.layer.Group

### LINKED ISSUE #3479: Memory Leak with adding and removing layers from a layer group.
I believe I have discovered a memory leak with adding and removing layers from a map. 

I produced this result in google chrome on Mac OSX
screen shot 2015-04-03 at 11 24 03 am

using the following html.

```
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="initial-scale=1.0, user-scalable=no, width=device-width">
    <link rel="stylesheet" href="../css/ol.css">
    <title>Add/Remove Layers example</title>
  </head>
  <body>
 
    <div class="container">
 
      <div class="row">
        <div class="md-12">
          <div id="map" class="map"></div>
        </div>
      </div>
      <div id="span12"> 
            <button id="add_layers">Add Layers</button>
            <button id="remove_layers">Remove All Layers</button>
            <label for="depth">
            <input id="depth" type="range" min="1" max="50" step="1" value="1"> &nbsp; # Layers:&nbsp;<span id="count">1</span>
            &nbsp;
          </label>
      </div>
    </div>
    <script src="../build/ol-debug.js"></script>
 
    <script>
                
var depthInput = document.getElementById('depth');
var num_layers = Number(depthInput.value);

var map = new ol.Map({
  layers: [],
  target: 'map',
  view: new ol.View({
    center: [0, 0],
    zoom: 2
  })
});
var ol_layer_group = new ol.layer.Group();
map.addLayer(ol_layer_group);



var add_layers_button = document.getElementById('add_layers');
add_layers_button.addEventListener('click', function() {
  // Add the given number of layers
  var num_layers = Number(depthInput.value);
  num_layers = num_layers * num_layers;
  var layers = ol_layer_group.getLayers();
  for (var j = 0; j < num_layers; j++) {
      var vector_layer = new ol.layer.Vector();
      layers.push(vector_layer);
  }
}, false);
var remove_layers_button = document.getElementById('remove_layers');
remove_layers_button.addEventListener('click', function() {
  // Remove all the layers
  var layers = ol_layer_group.getLayers();
  for (var j = 0; j < layers.getLength(); j++) {
      var layer = layers.pop();
      layer = {};
  }

}, false);
depthInput.onchange = function() {
    var num_layers = Number(depthInput.value);
    num_layers = num_layers * num_layers;
  document.getElementById('count').innerHTML = num_layers;

};
      
    </script>
 
  </body>
</html>

```

I also tested this using the non-debug version of ol and got very similar results. Any help that can be provided on this would be very useful. Thank you.

### PR DESCRIPTION
**Pull request by fredj**
Fixes #3479

### CODE REVIEW

File: src/ol/layer/layergroup.js [hunk 1]
```diff
@@ -123,9 +133,15 @@ ol.layer.Group.prototype.handleLayersChanged_ = function(event) {
  */
 ol.layer.Group.prototype.handleLayersAdd_ = function(collectionEvent) {
   var layer = /** @type {ol.layer.Base} */ (collectionEvent.element);
-  this.listenerKeys_[goog.getUid(layer).toString()] = goog.events.listen(
-      layer, [ol.ObjectEventType.PROPERTYCHANGE, goog.events.EventType.CHANGE],
-      this.handleLayerChange_, false, this);
+  var key = goog.getUid(layer).toString();
+  goog.asserts.assert(!goog.isDef(this.listenerKeys_[key]),
```
Comment (elemoine) [inline src/ol/layer/layergroup.js]: Using `goog.asserts.assert(!(key in this.listenerKeys_))` would make more sense to me.

File: src/ol/layer/layergroup.js [hunk 2]
```diff
@@ -89,28 +95,32 @@ ol.layer.Group.prototype.handleLayerChange_ = function() {
...
-        this.handleLayersAdd_, false, this),
-    'remove': goog.events.listen(layers, ol.CollectionEventType.REMOVE,
-        this.handleLayersRemove_, false, this)
-  };
+  this.layersListenerKeys_.push(
+      goog.events.listen(layers, ol.CollectionEventType.ADD,
+          this.handleLayersAdd_, false, this));
+  this.layersListenerKeys_.push(
+      goog.events.listen(layers, ol.CollectionEventType.REMOVE,
+          this.handleLayersRemove_, false, this));
```
Comment (elemoine) [inline src/ol/layer/layergroup.js]: You could just call `push` once here:

```js
this.layersListenerKeys_.push(
    goog.events.listen(layers, ol.CollectionEventType.ADD,
        this.handleLayersAdd_, false, this),
    goog.events.listen(layers, ol.CollectionEventType.REMOVE,
        this.handleLayersRemove_, false, this));
```

File: src/ol/layer/layergroup.js [hunk 3]
```diff
@@ -89,28 +95,32 @@ ol.layer.Group.prototype.handleLayerChange_ = function() {
...
+
+  goog.object.forEach(this.listenerKeys_, function(keys) {
+    goog.array.forEach(keys, goog.events.unlistenByKey);
+  });
+  goog.object.clear(this.listenerKeys_);
```
Comment (elemoine) [inline src/ol/layer/layergroup.js]: Do you think we should use `this.listenerKeys_ = {}` as you suggested it in https://github.com/openlayers/ol3/pull/3143? But I agree this discussion belongs to #3143…

**Comment by elemoine** [comment]
I've added a few minor comments. I'd also recommend adding unit-tests to check that that logic is correct and avoid regressions in the future.

File: src/ol/layer/layergroup.js (hunk 2 above)
Comment (fredj) [inline src/ol/layer/layergroup.js]: +1, thanks

**Comment by fredj** [comment]
@elemoine you're right; units tests must be added.

**Comment by fredj** [comment]
More unit tests added

**Comment by elemoine** [comment]
+1

**Pull request merged**